DataFormats/Common/interface/getThinned_implementation.h

```diff
@@ -46,61 +93,74 @@ namespace edm {
           continue;
         }
 
+        // Get the thinned container if it exists (need to check
+        // before the key because of constraints on slimmed
+        // containers
+        ProductID const& thinnedCollectionPID = thinnedAssociation->thinnedCollectionID();
+        WrapperBase const* thinnedCollection = getByProductID(thinnedCollectionPID);
+        if (thinnedCollection != nullptr and not associatedBranches->isSlimmed()) {
+          thinnedAvailable = true;
+        }
+
         // Does this thinned container have the element referenced by key?
         auto thinnedIndex = thinnedAssociation->getThinnedIndex(key);
         if (not thinnedIndex.has_value()) {
           continue;
```

Comment:
Does this consider the possibility that there might be a thinned collection that is a daughter of this thinnedCollection that is present and we should be setting thinnedAvailable to true even though we know that the collection cannot contain the element?

Reply:
By quick though I suspect you're right and this particular case is missed. I'll add a test for it. Thanks!

Reply:
Actually, given what I wrote in https://github.com/cms-sw/cmssw/pull/31100#issuecomment-675771110, the logic here is even more flawed. The decision whether to prefer a thinned collection over a slimmed one should be based on the existence of the branches in the job, not on the availability of the products (that may vary between events).

Reply:
> Does this consider the possibility that there might be a thinned collection that is a daughter of this thinnedCollection that is present and we should be setting thinnedAvailable to true even though we know that the collection cannot contain the element?

After adding a test this case was indeed not covered.

@wddgit If you have any further ideas of corner cases to be tested, I'm all ears.

Reply:
Done